harness/asyncHelpers.js

```diff
@@ -28,84 +28,88 @@ function asyncTest(testFunc) {
-        if (typeof thrown !== "object" || thrown === null) {
-          message += "Thrown value was not an object!";
+    try {
+      return innerThenable.then(
+        function () {
+          message +=
+            "Expected a " +
+            expectedErrorConstructor.name +
+            " to be thrown asynchronously but no exception was thrown at all";
           throw new Test262Error(message);
-        } else if (thrown.constructor !== expectedErrorConstructor) {
-          expectedName = expectedErrorConstructor.name;
-          actualName = thrown.constructor.name;
-          if (expectedName === actualName) {
-            message +=
-              "Expected a " +
-              expectedName +
-              " but got a different error constructor with the same name";
-          } else {
-            message +=
-              "Expected a " + expectedName + " but got a " + actualName;
+        },
+        function (thrown) {
+          var expectedName, actualName;
+          if (typeof thrown !== "object" || thrown === null) {
+            message += "Thrown value was not an object!";
+            throw new Test262Error(message);
+          } else if (thrown.constructor !== expectedErrorConstructor) {
+            expectedName = expectedErrorConstructor.name;
+            actualName = thrown.constructor.name;
+            if (expectedName === actualName) {
+              message +=
+                "Expected a " +
+                expectedName +
+                " but got a different error constructor with the same name";
+            } else {
+              message +=
+                "Expected a " + expectedName + " but got a " + actualName;
+            }
+            throw new Test262Error(message);
           }
-          throw new Test262Error(message);
         }
+      );
+    } catch (thrown) {
+      if (typeof thrown !== "object" || thrown === null) {
+        message +=
+          "Expected a " +
+          expectedErrorConstructor.name +
+          " to be thrown asynchronously but innerThenable synchronously threw a value that was not an object ";
+      } else {
+        message +=
+          "Expected a " +
+          expectedErrorConstructor.name +
+          " to be thrown asynchronously but a " +
+          thrown.constructor.name +
+          " was thrown synchronously";
       }
-    );
-  } catch (thrown) {
-    if (typeof thrown !== "object" || thrown === null) {
-      message +=
-        "Expected a " +
-        expectedErrorConstructor.name +
-        " to be thrown asynchronously but innerThenable synchronously threw a value that was not an object ";
-    } else {
-      message +=
-        "Expected a " +
-        expectedErrorConstructor.name +
-        " to be thrown asynchronously but a " +
-        thrown.constructor.name +
-        " was thrown synchronously";
+      throw new Test262Error(message);
     }
-    throw new Test262Error(message);
-  }
+
+    resolve();
```

Comment:
This line will never be reached; it is preceded by `try { return …; } catch (thrown) { …; throw new Test262Error(…); }`. I think you actually want to capture both `resolve` and `reject` in the `new Promise` callback, and replace `return innerThenable.then(…)` with `innerThenable.then(…).then(resolve, reject)`.
```patch
@@ -29,7 +29,7 @@ function asyncTest(testFunc) {
 }
 
 assert.throwsAsync = function (expectedErrorConstructor, func, message) {
-  return new Promise(function (resolve) {
+  return new Promise(function (resolve, reject) {
     var innerThenable;
     if (message === undefined) {
       message = "";
@@ -64,7 +64,7 @@ assert.throwsAsync = function (expectedErrorConstructor, func, message) {
     }
 
     try {
-      return innerThenable.then(
+      innerThenable.then(
         function () {
           message +=
             "Expected a " +
@@ -92,7 +92,7 @@ assert.throwsAsync = function (expectedErrorConstructor, func, message) {
             throw new Test262Error(message);
           }
         }
-      );
+      ).then(resolve, reject);
     } catch (thrown) {
       if (typeof thrown !== "object" || thrown === null) {
         message +=
@@ -109,7 +109,5 @@ assert.throwsAsync = function (expectedErrorConstructor, func, message) {
       }
       throw new Test262Error(message);
     }
-
-    resolve();
   });
 };

```